File: algorithm-design/leetcode/lc029_divide_two_integers/lc029_divide_two_integers.py
#!/usr/bin/env python3
# https://leetcode.com/problems/divide-two-integers/

import unittest


class Solution:
    MAX_VALUE = 2147483647

    def divide(self, dividend: int, divisor: int) -> int:
        if divisor == 0:
            return Solution.MAX_VALUE
        if dividend == 0:
            return 0
        sign = (dividend / abs(dividend)) * (divisor / abs(divisor))
        dividend = abs(dividend)
        divisor = abs(divisor)
        counter = 0
        while dividend >= divisor:
            subtractor = divisor
            incrementor = 1
            while dividend - subtractor - subtractor >= 0:
                subtractor += subtractor
                incrementor += incrementor
            dividend -= subtractor
            counter += incrementor
        if sign * counter > Solution.MAX_VALUE:
            return Solution.MAX_VALUE
        return sign * counter


class TestCode(unittest.TestCase):
    def test_Minus1010369383_Minus2147483648(self):
        self.assertEqual(
            Solution().divide(-1010369383, -2147483648),
            -1010369383 / -2147483648)

    def test_Minus2147483648_Minus1(self):
        self.assertEqual(
            Solution().divide(-2147483648, -1),
            Solution.MAX_VALUE)

    def test_divisor_0(self):
        self.assertEqual(
            Solution().divide(1, 0),
            Solution.MAX_VALUE)

    def test_dividend_0(self):
        self.assertEqual(
            Solution().divide(0, 1),
            0)

File: algorithms/code/leetcode/lc029_divide_two_integers/lc029_divide_two_integers.rs
// https://leetcode.com/problems/divide-two-integers/

use num_traits::abs;

pub struct Solution;

impl Solution {
    pub fn divide(dividend: i32, divisor: i32) -> i32 {
        if divisor == 0 {
            return i32::MAX;
        }
        if dividend == 0 {
            return 0;
        }
        let mut sign: i64 = 1;
        if (dividend >= 0 && divisor < 0) || (dividend < 0 && divisor >= 0) {
            sign = -1;
        }
        let mut dividend = abs(dividend as i64);
        let divisor = abs(divisor as i64);
        let mut counter: i64 = 0;
        while dividend >= divisor {
            let mut subtractor = divisor;
            let mut incrementor = 1;
            while dividend - subtractor - subtractor >= 0 {
                subtractor += subtractor;
                incrementor += incrementor;
            }
            dividend -= subtractor;
            counter += incrementor;
        }
        let result = sign * counter;
        if result > i32::MAX.into() {
            return i32::MAX;
        }
        result as i32
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_minus1010369383_minus2147483648() {
        assert_eq!(
            Solution::divide(-1010369383, i32::MIN),
            -1010369383 / i32::MIN
        );
    }

    #[test]
    fn test_minus2147483648_minus1() {
        assert_eq!(Solution::divide(i32::MIN, -1), i32::MAX);
    }

    #[test]
    fn test_divisor_0() {
        assert_eq!(Solution::divide(1, 0), i32::MAX);
    }

    #[test]
    fn test_dividend_0() {
        assert_eq!(Solution::divide(0, 1), 0);
    }
}
